design/API_requirements_and_design.md

```diff
@@ -0,0 +1,59 @@
+# Thin-Edge API requirements and design
+
+One should be able to build a thin-edge executable from tedge plugins crates that have been implemented independently:
+
+* All these tedge plugins crates depends on the tedge_api crate that defines how
+  to create, connect and run plugin instances.
+* Each tedge plugin crate can define its own types of messages. For instance, a
+  tedge_telemetry plugin can defines measurements, events, alarms and setpoints
+  messages.
+* To exchange typed messages, plugin crates need dependency relationship. For
+  instance, the tedge_json_over_mqtt plugin crate will have to depend on the
+  tedge_telemetry plugin crate and the tedge_mqtt crate.
+* The tedge_api crate defines only messages related to the runtime as Shutdown.
+* The implementation of the mechanisms defined by the tedge_api are provided by
+  a tedge_runtime crate.
+* The final executable is build as an assemblage of all these plugin crates and
+  a `main.rs` that defines which plugins have to be instantiated and then
+  launched.
+
+---------
+
+# Non-functional Requirements
+
+- Plugins can be included in the final executable without having them 'instantiated'
+  - The motivation is to be able to deliver a battery included executable, ready
+    to be used for a large set of use-cases without any Rust expertise.
+  - Notably a battery-included executable should include an MQTT connection
+    plugin and a JSON plugin for telemetry messages.
+- Plugins are instantiated and configured through an external configuration at
+  runtime
+- At startup the configuration is read and from it a set of plugins 'to be
+  instantiated' is derived
+- One should be able to deduplicate parts of the configuration (to prevent
+  repetition)
+- Should plugin instances be changeable during runtime? (open question)
+- Should some plugin configurations be changeable at runtime or not?
+- Having an optional plugin handling SIGHUP and send an internal message
+  UpdateConfig to all the running plugin instances.
+- Would having multiple configuration files be an option? (open question)
```

Comment:
I think we need to put apart two aspects and to make this separation explicit with two kind of configuration files:

* Plugin instantiation and inter-connection. This defines the application with its set of features.
* Runtime configuration (IP addresses, external resources, internal capacity tuning ...). This defines the deployment of the application in a given operational context.

One should be able to update one aspect without risks to break the other. Furthermore, these two aspects are defined by different persons with different schedules.

From my point of view, a configuration akin to [this example](https://github.com/matthiasbeyer/thin-edge.io/blob/feature/add_tedge_api_impl_typed/tedge/example-config.toml) mixes the two aspects and will be hard to maintain for a realistic application.

Reply:
I suppose one way to get out of this would be to have generic configuration (as it is proposed now) but allow for inclusion of multiple files, this could be achieved by adding a clause in config `include` which would be a path allowing for pointing to a file or directory and leave to the user to decide if they prefer to have everything in a single file or multiple ones. 
The configuration reading phase would then collect all the configs, validate them and load.